### PR TITLE
IGNITE-17291

### PR DESCRIPTION
**Pull request by denis-chudov**

### CODE REVIEW

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java [hunk 1]
```diff
@@ -50,19 +52,41 @@
...
+     * @param fn                    Function transforming the element of type {@link M} to the type of {@link T}.
+     */
+    public <M> CursorImpl(
+            RaftGroupService metaStorageRaftGrpSvc,
+            CompletableFuture<IgniteUuid> initOp,
+            Function<M, T> fn
+    ) {
+        this.metaStorageRaftGrpSvc = metaStorageRaftGrpSvc;
+        this.initOp = initOp;
+        this.it = new InnerIterator(e -> singleton(e), fn);
```
Comment (sanpwc) [inline modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java:69]: Do you really like an idea of wrapping every next result with SingletonSet?

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java [hunk 2]
```diff
@@ -50,19 +52,41 @@
...
      * @param initOp                Future that runs meta storage service operation that provides cursor.
+     * @param internalCacheFn       Function transforming the result of {@link CursorNextCommand} to the
+     *                              {@link Iterable} of elements of type {@link M}.
+     * @param fn                    Function transforming the element of type {@link M} to the type of {@link T}.
      */
-    CursorImpl(RaftGroupService metaStorageRaftGrpSvc, CompletableFuture<IgniteUuid> initOp, Function<Object, T> fn) {
+    public <M> CursorImpl(
+            RaftGroupService metaStorageRaftGrpSvc,
+            CompletableFuture<IgniteUuid> initOp,
+            Function<Object, Iterable<M>> internalCacheFn,
```
Comment (sanpwc) [inline modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java:84]: I don't think that we need two explicit transformation functions. Please consider approach that is used within
`new CursorImpl<>(metaStorageRaftGrpSvc, watchRes, MetaStorageServiceImpl::watchResponse),`.

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/EntryImpl.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public final class EntryImpl implements Entry {
      * @param rev     Revision.
      * @param updCntr Update counter.
      */
-    EntryImpl(@NotNull ByteArray key, @Nullable byte[] val, long rev, long updCntr) {
+    public EntryImpl(@NotNull ByteArray key, @Nullable byte[] val, long rev, long updCntr) {
```
Comment (sanpwc) [inline modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/EntryImpl.java:51]: package-private still seems fine here.

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/MetaStorageRangeCursorTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+/**
+ * Test for cursor created by {@link MetaStorageService#range(ByteArray, ByteArray)} commands.
+ */
+@ExtendWith(MockitoExtension.class)
+public class MetaStorageRangeCursorTest {
```
Comment (sanpwc) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/MetaStorageRangeCursorTest.java:66]: I'd rather move given range test to org.apache.ignite.internal.metastorage.client.ItMetaStorageServiceTest, WDYT?

Comment (denis-chudov) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/MetaStorageRangeCursorTest.java:66]: `ItMetaStorageServiceTest` are more slow due to `@BeforeEach` method, this is unacceptable for the tests in `MetaStorageRangeCursorTest` that has several dozens of tests.

**Pull request merged**